daffodil-runtime1/src/main/scala/org/apache/daffodil/layers/LayerTransformer.scala

```diff
@@ -282,4 +234,234 @@ case class LayerNotEnoughDataException(layerSRD: SequenceRuntimeData, state: PSt
   extends LayerException(layerSRD, state, One(cause), None, nBytesRequired) {
   override def isError = true
   override def modeName = "Parse"
-  }
\ No newline at end of file
+  }
+
+
+
+class LayerRuntimeInfoImpl(val srd: SequenceRuntimeData,
+  maybeLayerCharsetEv: Maybe[LayerCharsetEv],
+  maybeLayerLengthKind: Maybe[LayerLengthKind],
+  maybeLayerLengthEv: Maybe[LayerLengthEv],
+  maybeLayerLengthUnits: Maybe[LayerLengthUnits],
+  maybeLayerBoundaryMarkEv: Maybe[LayerBoundaryMarkEv])
+  extends LayerRuntimeInfo {
+
+  def evaluatables: Seq[Evaluatable[AnyRef]] =
+    maybeLayerCharsetEv.toScalaOption.toSeq ++
+      maybeLayerLengthEv.toScalaOption.toSeq ++
+      maybeLayerBoundaryMarkEv.toScalaOption.toSeq
+
+
+  def layerInfo(state: ParseOrUnparseState): LayerInfo =
+    new LayerInfoImpl(state, srd,
+      maybeLayerCharsetEv,
+      maybeLayerLengthKind,
+      maybeLayerLengthEv,
+      maybeLayerLengthUnits,
+      maybeLayerBoundaryMarkEv)
+}
+
+/**
+ * Allows access to all the layer properties, if defined, including
+ * evaluating expressions if the properties values are defined as expressions.
+ */
+sealed trait LayerInfo {
```

Comment:
It's not totally clear to me the difference between `LayerCompileInfo`, `LayerRuntimeInfo`, and `LayerInfo`. Is this correct:

`LayerCompileInfo` contains information that can be used to determine what `LayerFactory` is created, but should not be serialized?

`LayerRuntimeInfo` can contain whatever information the Layer needs to do it's thing. The default implemetation is basically the same as the `LayerCompileInfo`, but this need not be the case. It is serialized along with the `LayerTransformFactory` and can be used to determine how a `LayerTransformer` is created at parse/unparse time when `newInstance` is called.

`LayerInfo` only makes sense when `LayerRuntimeInfo` is the specific `Impl` kind, and contains the result of expressions that couldn't be evaluated at compile time. The results of these expression cannot be used to determine the parameters of the LayerTransformer--it only can affect the wrapLimitingStream function.

Is this accurate? If so, it's not clear to me why we need these three things. Could we have `LayerCompileInfo` as is, and it is only used to create the `LayerTransfomerFactory`. Then at parse time, we create LayerInfo (suggest renaming to LayerRuntimeInfo, and pass that into the `newInstance`. Now that instance has all the evaluated runtime values, and can create the right instance, and use it for things like wrapping, etc. This way the concept of a Runtime thing being stored in the serialized Factory doesn't exist. Runtime inforation is only available for Transformer.

Reply:
So, LayerCompileInfo is everything needed to compile the layer - and issue errors at compile time. 

LayerRuntimeInfo isn't everything the layer needs. It is just a serialization of the layer-oriented property values that were defined for that layer. These are already known to be correct and adequate based on LayerCompileInfo. 
Anything else the layer needs, by way of parameters, has to be supplied via DFDL variables at runtime. 

The reason we need LayerIinfo, is that there are things that cannot be computed without a PState/UState because some properties can be expressions. LayerInfo contains the result of evaluating those evaluatables given the PState/UState passed at runtime to the transformation methods. This includes the layer length (from LayerLengthEv), the LayerBoundaryMarkEv, and I think the layer encoding as well. 

But you make a good point which is that we are already doing the separate compile step now, so perhaps the LayerTransformerFactory becomes internal, and the parser/unparser creates the LayerTransformer instance the same way every time, and does so only once we have a PState/UState, at real parse/unparse time.

I will look into this, as the 3 level cake here does seem like 1 too many.